front/src/modules/simulationResult/components/ChartHelpers/enableInteractivity.tsx

```diff
@@ -0,0 +1,384 @@
+              : timePosition
+          )
+        );
+    } else {
+      chart.svg
+        .selectAll('#vertical-line')
+        .attr(
+          'x1',
+          chart.x(
+            !isSpaceTimeChart(keyValues) && positionValues.speed
+              ? positionValues.speed.position
+              : timePosition
+          )
+        )
+        .attr(
+          'x2',
+          chart.x(
+            !isSpaceTimeChart(keyValues) && positionValues.speed
+              ? positionValues.speed.position
+              : timePosition
+          )
+        );
+    }
+
+    updatePointers(chart, keyValues, positionValues, rotate);
+  }
+};
+
+// enableInteractivity
+
+// Override the default wheelDelta computation to get smoother zoom
+function wheelDelta(event: WheelEvent) {
+  let factor = 1;
+  if (event.deltaMode === 1) {
+    factor = 0.005;
+  } else if (event.deltaMode) {
+    factor = 0.01;
+  } else {
+    factor = 0.002;
+  }
+
+  return -event.deltaY * factor;
+}
+
+export const enableInteractivity = <
+  T extends Chart | SpeedSpaceChart,
+  Data extends SimulationTrain<Date> | GevPreparedData | SpaceCurvesSlopesData
+>(
+  chart: T | undefined,
+  selectedTrainData: Data,
+  keyValues: ChartAxes,
+  rotate: boolean,
+  setChart: React.Dispatch<React.SetStateAction<T | undefined>>,
+  simulationIsPlaying: boolean,
+  dispatchUpdateTimePositionValues: (newTimePositionValues: Date) => void,
+  chartDimensions: [Date, Date]
+) => {
+  if (!chart) return;
+
+  const zoom = d3zoom<SVGGElement, unknown>()
+    .scaleExtent([0.3, 20]) // This controls how much you can unzoom (x0.5) and zoom (x20)
+    .extent([
+      [0, 0],
+      [chart.width, chart.height],
+    ])
+    .wheelDelta(wheelDelta)
+    .on('zoom', (event) => {
+      event.sourceEvent.preventDefault();
+      const zoomFunctions = updateChart(chart, keyValues, rotate, event);
+      const newChart = { ...chart, x: zoomFunctions.newX, y: zoomFunctions.newY };
+      setChart(newChart);
+    })
+    .filter(
+      (event) => (event.button === 0 || event.button === 1) && (event.ctrlKey || event.shiftKey)
+    );
+
+  // TODO: actually a number as we’re not in node.js
+  let debounceTimeoutId: NodeJS.Timeout;
+  function debounceUpdateTimePositionValues(timePositionLocal: Date, interval: number) {
+    clearTimeout(debounceTimeoutId);
+    debounceTimeoutId = setTimeout(() => {
+      dispatchUpdateTimePositionValues(timePositionLocal);
+    }, interval);
+  }
+
+  const mousemove = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+    if (!simulationIsPlaying) {
+      let immediatePositionsValuesForPointer: ReturnType<typeof interpolateOnTime>;
+      let timePositionLocal: Date | null;
+      if (isSpaceTimeChart(keyValues)) {
+        // SpaceTimeChart
+        timePositionLocal = (
+          rotate
+            ? chart.y.invert(pointer(event, event.currentTarget)[1])
+            : chart.x.invert(pointer(event, event.currentTarget)[0])
+        ) as Date;
```

Comment:
Remove the cast here and add a test after this if/else block
```suggestion
      );
```

Reply:
interpolateOnTime only wants `Date`

Reply:
interpolateOnTime only wants Date.

and i think it makes sense to make that cast. once we’re inside the if we know timePositionLocal is defined because we do the assign. and we make sure

Reply:
Yes you're right :+1: